logstash-core/lib/logstash/pipeline.rb

```diff
@@ -100,7 +100,9 @@ def initialize(config_str, settings = SETTINGS, namespaced_metric = nil)
       raise
     end
     @queue = build_queue_from_settings
-    @input_queue_client = @queue.write_client
+
+    @input_queue_client_factory = LogStash::Instrument::WriteClientFactory.new(queue.write_client, self, metric)
```

Comment:
We use the same pattern in the LIR execution branch to encapsulate all the metric logic this is why I went that route with the current ruby implementation.